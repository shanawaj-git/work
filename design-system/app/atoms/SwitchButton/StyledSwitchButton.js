import styled from 'styled-components';
import { Switch } from '@mui/material';
import React from 'react';
import { colorSelector } from 'themes';

const pxToRem = (px, oneRemPx = 17) => `${px / oneRemPx}rem`;

const borderWidth = 2;
const size = pxToRem(20.5);
const gap = (34 - 22) / 2;

export const StyledSwitchButton = styled(props => <Switch {...props} />)(
  () => ({
    color: colorSelector('c200'),
    '.MuiSwitch-track': {
      position: 'absolute',
      opacity: `1`,
      borderRadius: '20px',
      borderWidth,
      backgroundColor: colorSelector('c200'),
      border: `1px solid ${colorSelector('g100')}`,
      boxSizing: 'border-box',
      top: '9%',
      left: '1px',
      height: '65%',
      width: '80%',
    },
    '& .MuiSwitch-thumb': {
      boxShadow: 'none',
      backgroundColor: colorSelector('c100'),
      width: size,
      height: size,
    },
    '& .MuiSwitch-switchBase': {
      padding: pxToRem(gap),
      left: '-2px',
    },

    '& .Mui-checked ': {
      '.MuiSwitch-thumb': {
        backgroundColor: colorSelector('c200'),
      },
    },

    '&& .Mui-checked+.MuiSwitch-track': {
      opacity: 1,
      backgroundColor: colorSelector('c100'),
    },

    '& .Mui-disabled+.MuiSwitch-track': {
      background: colorSelector('g100'),
      opacity: `${1}!important`,
      border: 'none',
    },

    '& .Mui-disabled': {
      '>.MuiSwitch-thumb': {
        backgroundColor: colorSelector('g300'),
      },
    },
  }),
);
