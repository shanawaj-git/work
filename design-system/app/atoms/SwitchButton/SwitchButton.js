import React from 'react';
import PropTypes from 'prop-types';

import { StyledSwitchButton } from './StyledSwitchButton';

const SwitchButton = props => {
  const { onChange, checked = false, disabled = false, others } = props;
  const [toggled, setToggled] = React.useState(checked);

  const toggle = event => {
    const isChecked = event.target.checked;
    // update the UI
    setToggled(isChecked);
    // execute onChange function passed in props
    onChange(toggled);
  };
  return (
    <div>
      <StyledSwitchButton
        checked={toggled}
        disabled={disabled}
        onChange={toggle}
        {...others}
        data-testid="toggle"
      />
    </div>
  );
};

SwitchButton.propTypes = {
  checked: PropTypes.bool,
  disabled: PropTypes.bool,
  onChange: PropTypes.func.isRequired,
  others: PropTypes.object,
};

export default SwitchButton;
