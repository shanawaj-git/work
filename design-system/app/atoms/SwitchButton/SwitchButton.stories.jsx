import React from 'react';

import SwitchButton  from './SwitchButton';
import StoryWrapper from 'molecules/utility/StoryWrapper';

export default {
  title: 'Components/Buttons/SwitchButton',
  component: SwitchButton,
  decorators: [
    (Story) => (
		<StoryWrapper>
			{Story()}
		</StoryWrapper>
    ),
  ],
};

const Template = (args) => <SwitchButton  {...args} />;

export const Unchecked = Template.bind({});
Unchecked.args = {
  disabled: false,
};

export const Checked = Template.bind({});
Checked.args = {
  checked: true,
  disabled: false,
};

export const DisabledChecked = Template.bind({});
DisabledChecked.args = {
  checked: true,
  disabled: true,
};

export const DisabledUnchecked = Template.bind({});
DisabledUnchecked.args = {
  checked: false,
  disabled: true,
};