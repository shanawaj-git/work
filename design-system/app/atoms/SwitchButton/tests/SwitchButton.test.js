import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import renderer, { act } from 'react-test-renderer';
import { render, unmountComponentAtNode } from 'react-dom';

import SwitchButton from '../index';

const shallowRenderer = new ShallowRenderer();
let container = null;

describe('SwjtchButton.js', () => {
  beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    // cleanup on exiting
    unmountComponentAtNode(container);
    container.remove();
    container = null;
  });

  describe('index', () => {
    it('Should import succesfully', () => {
      const tree = renderer.create(<SwitchButton onChange={jest.fn()} />);
      expect(tree).toBeTruthy();
    });

    it('Should display a checked switch button', () => {
      shallowRenderer.render(<SwitchButton checked onChange={jest.fn()} />);
      const renderedOutput = shallowRenderer.getRenderOutput();

      expect(renderedOutput).toMatchSnapshot();
    });

    it('Should display a disabled switch button', () => {
      shallowRenderer.render(<SwitchButton disabled onChange={jest.fn()} />);
      const renderedOutput = shallowRenderer.getRenderOutput();

      expect(renderedOutput).toMatchSnapshot();
    });

    it('should call onChange function in case of onchange event happens', () => {
      const onChangeMock = jest.fn();
      const testRenderer = renderer.create(
        <SwitchButton onChange={onChangeMock} />,
      );
      const testInstance = testRenderer.root;
      const mEvent = { target: { value: 'checked' } };
      act(() => {
        testInstance.findByType(SwitchButton).props.onChange(mEvent);
      });
      expect(testRenderer).toMatchSnapshot();
      expect(onChangeMock).toHaveBeenCalledTimes(1);
      expect(onChangeMock).toHaveBeenCalledWith(mEvent);
    });

    it('changes value when clicked', () => {
      const onChange = jest.fn();
      act(() => {
        render(<SwitchButton onChange={onChange} />, container);
      });

      // get a hold of the input element, and trigger some clicks on it
      const input = document.querySelector('[data-testid=toggle] input');

      act(() => {
        input.dispatchEvent(new MouseEvent('click', { bubbles: true }));
      });

      expect(onChange).toHaveBeenCalledTimes(1);

      act(() => {
        for (let i = 0; i < 5; i += 1) {
          input.dispatchEvent(new MouseEvent('click', { bubbles: true }));
        }
      });

      expect(onChange).toHaveBeenCalledTimes(6);
    });
  });
});
