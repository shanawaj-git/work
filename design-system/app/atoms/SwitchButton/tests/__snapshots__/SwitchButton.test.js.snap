// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwjtchButton.js index Should display a checked switch button 1`] = `
<div>
  <ForwardRef(Styled(Component))
    checked={true}
    data-testid="toggle"
    disabled={false}
    onChange={[Function]}
  />
</div>
`;

exports[`SwjtchButton.js index Should display a disabled switch button 1`] = `
<div>
  <ForwardRef(Styled(Component))
    checked={true}
    data-testid="toggle"
    disabled={true}
    onChange={[Function]}
  />
</div>
`;

exports[`SwjtchButton.js index should call onChange function in case of onchange event happens 1`] = `
<div>
  <span
    className="MuiSwitch-root MuiSwitch-sizeMedium sc-bdVaJa fimhKu css-julti5-MuiSwitch-root"
  >
    <span
      className="MuiSwitch-switchBase MuiSwitch-colorPrimary MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-root css-5ryogn-MuiButtonBase-root-MuiSwitch-switchBase"
      data-testid="toggle"
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={null}
    >
      <input
        checked={false}
        className="MuiSwitch-input PrivateSwitchBase-input css-1m9pwf3"
        disabled={false}
        onChange={[Function]}
        type="checkbox"
      />
      <span
        className="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
      />
      <span
        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </span>
    <span
      className="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
    />
  </span>
</div>
`;
